keps/sig-storage/361-local-ephemeral-storage-isolation/README.md

```diff
@@ -0,0 +1,482 @@
+  container level. So similar to CPU and memory, container spec can specify the request/limit
+  for local storage. All the validation related to the storage resource requirements will be the
+  same as memory resource. In the case of two partition, only the storage usage in root partition
+  will be isolated across different containers. (In another word, second or more partitions
+  will not be counted for container-level eviction management)
+* Pod-level resource constraint: Since we haven’t supported pod-level resource API, the sum of the
+  resource request/limit is considered as pod-level resource requirement. Similar to memory,
+  all local ephemeral storage resource usage is subject to this requirement. For example,
+  emptyDir disk usage (also secrets, configuMap, downwardAPI, gitRepo since they are wrapped
+  emptyDir volume) plus containers disk usage should not exceed the pod-level resource requirement
+  (the sum of all container’s limit). 
+* EmptyDir SizeLimit: Because emptyDir volume is a pod-level resource which is managed separately,
+  we also add a sizeLimit for emptyDir Volume for additional storage isolation. If this limit is
+  set for emptyDir volume (default medium), eviction manager will validate this limit with emptyDir
+  usage too in additional to the above pod-level resource constraints. 
+
+### Eviction Policy and Scheduler Predicates
+
+CPU and memory use cgroup for limiting the resource usage. However, cgroup for disk usage is not
+available. Our current design is to evict pods when exceeding the limit set for local storage.
+The eviction policy is listed as follows.
+
+* If the container writable layer (overlay) usage exceeds its limit set for this container, pod gets evicted.
+* If the emptyDir volume usage exceeds the sizeLimit, pod gets evicted.
+* If the sum of the usage from emptyDir and all contains exceeds the sum of the container’s local storage
+  limits (pod-level limit), pod gets evicted.
+
+Pod Admission
+
+* When scheduler admits pods, it sums up the storage requests from each container and pod can be scheduled
+  only if the sum is smaller than the allocatable of the local storage space.
+
+Resource Quota
+
+This feature adds two more resource quotas for storage. The request and limit set constraints on the
+total requests/limits of all containers’ in a namespace.
+
+* requests.ephemeral-storage
+* limits.ephemeral-storage
+
+LimitRange 
+
+* Similar to CPU and memory, admin could use LimitRange to set default container’s local storage
+  request/limit, and/or minimum/maximum resource constraints for a namespace. 
+
+Node Allocatable Resources
+
+* Similar to CPU and memory, ephemeral-storage may be specified to reserve for kubelet or system.
+  example,
+  --system-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]
+  --kube-reserved=[cpu=100m][,][memory=100Mi][,][ephemeral-storage=1Gi][,][pid=1000]
+
+### Risks and Mitigations
```

Comment:
Much time has passed since beta. Now we also have CSI ephemeral volumes which can be another source of local space consumption. Can we think about what kind of impact it has to this feature?

Reply:
Updated based on our offline conversation.